jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java

```diff
@@ -397,6 +407,31 @@ public boolean isEmpty()
      */
     public void deleteParts()
     {
+        while (true)
+        {
+            switch (state.get())
+            {
+                case PARSING:
+                    state.compareAndSet(State.PARSING, State.ERROR);
+                    Thread.yield();
+                    continue;
+
+                case UNPARSED:
+                    if (!state.compareAndSet(State.UNPARSED, State.COMPLETED))
+                        continue;
+                    break;
+
+                case ERROR:
```

Comment:
Maybe both ERROR and COMPLETED are both terminal states?

Reply:
Just moving to `ERROR` state is not enough, the `ERROR` state is indicating that the parsing thread should exit, and the move to `COMPLETED` indicates that it has exited and will not create anymore parts.